brave/brave-core
# Asset discovery

Resolves https://github.com/brave/brave-browser/issues/19746.
Security Review https://github.com/brave/security/issues/1003

This adds logic that automatically detect and add ERC20 tokens on Mainnet Ethereum to the the list of user assets when new accounts are added.

## How it works
We make an eth_getLogs RPC request that returns all events that
* Were created between the "earliest" and "latest" blocks (all blocks). Note: scanning all blocks is intensive and requires the RPC provider to have an efficient eth_getLogs implementation like [Infura](https://blog.infura.io/post/faster-logs-and-events-e43e2fa13773).
* Match the ERC20 Transfer event signature
* Match the user's account address for the "to address" of the event
* Are emitted by ERC20 token contracts whose addresses are in our blockchain registry.

Each distinct address in the returned logs is an ERC20 contract of a token the user held at some point in the past. 

## Implementation (out of date, to be updated)
* Modify KeyringService to emit a AccountsAdded to be consumed by KeyringServiceObservers
* Add new AssetDiscoveryService class that implements KeyringServiceObserver::AccountsAdded which 
   * Filters out any added account that is not CoinType::ETH
   * Calls JsonRpcService::DiscoverAssets
* Add JsonRpcService::DiscoverAssets which 
    * Creates a list of contract_addresses to look for transfer events for by
        * Fetching all supported mainnet Ethereum tokens from the BlockchainRegistry
        * Removing any tokens that the user has already added
        * Removing any non ERC20 tokens
    * Creates a list of topics for the eth_getLogs call that will match all ERC20 transfer events to or from the any of the added accounts that were emitted by token contract addresses from the registry (eth_topics_builder.cc)
    * Makes the eth_getLogs RPC call, parses the returned logs, dedupes the list of contract addresses associated with discovered assets and returns the list of assets
* Adds the discovered assets by calling BraveWalletService::AddUserAsset for each
## Submitter Checklist:


## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:
Set up an account on Ethereum mainnet that has received an ERC20 transfer from an asset that is in our token registry, but are not added as visible assets.  Add this account using each of these methods, and verify the asset is now a visible asset:
* Create wallet
* Restore wallet
* Add account (standard way)
* Import private key
* Import hardware account

Some of these methods only work if the account that received the transfer is the first account derived.
